### PR TITLE
Additional roles and privileges APIs customization

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR folds together the following:

* Support fetching native-only roles, i.e., excluding reserved roles for the Get Roles API
* Switch the Delete Roles API to public protection scope
* Support injecting a response translator for the Get Builtin Privileges API

Depends on: https://github.com/elastic/elasticsearch/pull/105336

Relates: ES-7826, ES-7828, ES-7845

### CODE REVIEW

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/GetBuiltinPrivilegesRequest.java [hunk 1]
```diff
@@ -8,19 +8,12 @@
...
-import org.elasticsearch.common.io.stream.StreamInput;
-
-import java.io.IOException;
 
 /**
  * Request to retrieve built-in (cluster/index) privileges.
  */
 public final class GetBuiltinPrivilegesRequest extends ActionRequest {
 
-    public GetBuiltinPrivilegesRequest(StreamInput in) throws IOException {
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/GetBuiltinPrivilegesRequest.java:20]: This is a local-only API; porting it over since I'm already touching this code.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/role/GetRolesRequest.java [hunk 2]
```diff
@@ -47,9 +44,16 @@ public String[] names() {
...
+        this.nativeOnly = nativeOnly;
+    }
+
+    public boolean nativeOnly() {
+        return this.nativeOnly;
+    }
+
     @Override
     public void writeTo(StreamOutput out) throws IOException {
-        super.writeTo(out);
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/role/GetRolesRequest.java:52]: Likewise local-only -- I'm adding a new field so porting this over to avoid adding dead serialization code.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java [hunk 3]
```diff
@@ -27,13 +33,21 @@
 import static org.elasticsearch.rest.RestRequest.Method.GET;
 
 /**
- * Rest action to retrieve an application privilege from the security index
+ * Rest action to retrieve built-in (cluster/index) privileges
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java:36]: Drive-by fix.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java [hunk 4]
```diff
@@ -48,20 +62,45 @@ public String getName() {
...
     }
 
+    @Override
+    protected Exception innerCheckFeatureAvailable(RestRequest request) {
+        final boolean restrictPath = request.hasParam(RestRequest.RESPONSE_RESTRICTED);
+        assert false == restrictPath || DiscoveryNode.isStateless(settings);
+        if (false == restrictPath) {
+            return super.innerCheckFeatureAvailable(request);
+        }
+        // This is a temporary hack: we are re-using the native roles setting as an overall feature flag for custom roles.
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java:90]: Let me know if you feel this is too hacky -- could add a new system property to hide this API instead.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/role/RestGetRolesAction.java [hunk 5]
```diff
@@ -57,25 +54,43 @@ public String getName() {
...
-            public RestResponse buildResponse(GetRolesResponse response, XContentBuilder builder) throws Exception {
-                builder.startObject();
-                for (RoleDescriptor role : response.roles()) {
-                    builder.field(role.getName(), role);
-                }
-                builder.endObject();
+        final boolean restrictRequest = request.hasParam(RestRequest.RESPONSE_RESTRICTED);
+        assert false == restrictRequest || DiscoveryNode.isStateless(settings);
+        return channel -> new GetRolesRequestBuilder(client).names(roles)
+            .nativeOnly(restrictRequest)
```
Comment (n1v0lg) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/role/RestGetRolesAction.java:60]: We could consider exposing a `native_only` (or similar) request parameter publicly. In stateful, users could freely set this. In stateless, we'd restrict the ability to set it to operator users. Don't think we need to worry about this in the initial iteration though.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java [hunk 6]
```diff
@@ -48,20 +62,45 @@ public String getName() {
...
+        final Boolean nativeRolesEnabled = settings.getAsBoolean(NativeRolesStore.NATIVE_ROLES_ENABLED, true);
+        if (nativeRolesEnabled == false) {
```
Comment (slobodanadamovic) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java:92]: nit: Even if this is temporary, it might make sense to extract this code to a dedicated util method in order to avoid code duplication.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java [hunk 7]
```diff
@@ -48,20 +62,45 @@ public String getName() {
...
                     builder.endObject();
                     return new RestResponse(RestStatus.OK, builder);
                 }
             }
         );
     }
 
+    @Override
+    protected Exception innerCheckFeatureAvailable(RestRequest request) {
+        final boolean restrictPath = request.hasParam(RestRequest.RESPONSE_RESTRICTED);
```
Comment (slobodanadamovic) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java:85]:
```suggestion
        final boolean restrictPath = request.hasParam(RestRequest.PATH_RESTRICTED);
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java [hunk 8]
```diff
@@ -48,20 +62,45 @@ public String getName() {
 
     @Override
     public RestChannelConsumer innerPrepareRequest(RestRequest request, NodeClient client) throws IOException {
+        final boolean restrictResponse = request.hasParam(RestRequest.RESPONSE_RESTRICTED);
```
Comment (slobodanadamovic) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java:65]: Should we reference here the `PATH_RESTRICTED` instead?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/role/RestGetRolesAction.java (hunk 5 above)
Comment (slobodanadamovic) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/role/RestGetRolesAction.java:59]: Agreed on usefulness. Let's expose it if there is a need for it.

**Review by slobodanadamovic — APPROVED** [review]
LGTM 👍 

Left a question and some optional comments that should not require another round of review.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java (hunk 6 above)
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java:92]: I considered this but the other place it's used is in `NativeRoleBaseRestHandler` -- to share it between that and the privileges action would require a static method that takes multiple parameters (including the logger) -- given that it's essentially a simple if-else with a log I'm leaning towards simply sticking with the duplication.

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@elasticmachine update branch

**Comment by n1v0lg** [comment]
@elasticmachine update branch
